From 683fe6663a4b9b1ce350a93dab36dbe7a6e2ce5f Mon Sep 17 00:00:00 2001
From: Chris Renshaw <osm0sis@outlook.com>
Date: Sat, 2 Oct 2021 01:11:12 -0300
Subject: [PATCH 18/34] fix hush: add missing definitions on Android API 21+

platform FAST_FUNC adjustment to support NDK unified headers by John Wu <topjohnwu@gmail.com>

issetugid is required for hush via glob.c; add the hacky pre-API 21 definition back

Rebased for busybox 1.34.1 by Chris Renshaw <osm0sis@outlook.com>
---
 include/platform.h | 10 +++++++++-
 libbb/platform.c   |  8 ++++++++
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/include/platform.h b/include/platform.h
index 668b50df4..8b7ed16e2 100644
--- a/include/platform.h
+++ b/include/platform.h
@@ -124,7 +124,7 @@
  * (-DFAST_FUNC= )
  */
 #ifndef FAST_FUNC
-# if __GNUC_PREREQ(3,0) && defined(i386)
+# if !defined(__ANDROID__) && __GNUC_PREREQ(3,0) && defined(i386)
 /* stdcall makes callee to pop arguments from stack, not caller */
 #  define FAST_FUNC __attribute__((regparm(3),stdcall))
 /* #elif ... - add your favorite arch today! */
@@ -426,6 +426,7 @@ typedef unsigned smalluint;
 #define HAVE_SYS_STATFS_H 1
 #define HAVE_PRINTF_PERCENTM 1
 #define HAVE_WAIT3 1
+#define HAVE_ISSETUGID 1
 #define HAVE_DEV_FD 1
 #define DEV_FD_PREFIX "/dev/fd/"
 
@@ -533,6 +534,9 @@ typedef unsigned smalluint;
 # else
    /* ANDROID >= 21 has standard dprintf */
 # endif
+# if __ANDROID_API__ > 18
+#  undef HAVE_ISSETUGID
+# endif
 # if __ANDROID_API__ < 21
 #  undef HAVE_TTYNAME_R
 #  undef HAVE_GETLINE
@@ -633,4 +637,8 @@ extern int vasprintf(char **string_ptr, const char *format, va_list p) FAST_FUNC
 extern ssize_t getline(char **lineptr, size_t *n, FILE *stream) FAST_FUNC;
 #endif
 
+#ifndef HAVE_ISSETUGID
+extern int issetugid(void) FAST_FUNC;
+#endif
+
 #endif
diff --git a/libbb/platform.c b/libbb/platform.c
index 7913353e2..b652a2112 100644
--- a/libbb/platform.c
+++ b/libbb/platform.c
@@ -225,3 +225,11 @@ int ttyname_r(int fd, char *buf, size_t buflen)
 	return 0;
 }
 #endif
+
+#ifndef HAVE_ISSETUGID
+int issetugid(void)
+{
+	/* for Bionic, this is sufficient */
+	return 0;
+}
+#endif
-- 
2.31.1

